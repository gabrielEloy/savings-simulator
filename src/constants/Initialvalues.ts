export const FUTURE_MONTHS = 10;
export const MONEY_GOAL = 25_000;