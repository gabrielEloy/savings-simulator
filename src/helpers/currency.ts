import INumberCurrencyToString from 'interfaces/INumberCurrencyToString';

const removeString = (value: string, removedString: string) => {
    let  newValue = '';

    for(let char of value){
        if(char !==  removedString){
            newValue += char;
        }
    }

    return newValue;
}

export const stringCurrenctyToNumber = (value: string, options?: any) => parseFloat(removeString(value, ','));

export const numberCurrencyToString = (value: number = 0, options: INumberCurrencyToString = {}) => {
        const { targetLanguage = 'en-US' }  = options;
    
    const localization = new Intl.NumberFormat(targetLanguage, { style: 'currency', currency: 'usd', ...options});
    return localization.format(value)
}