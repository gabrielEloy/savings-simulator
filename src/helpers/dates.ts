export const getStringDate = (monthsAhead: number) => {
  const futureDate = new Date().setMonth(new Date().getMonth() + (monthsAhead - 1));

  const stringDate = new Intl.DateTimeFormat("en-US", {
    month: "long",
    year: "numeric",
  }).format(futureDate);

  return stringDate;
};
