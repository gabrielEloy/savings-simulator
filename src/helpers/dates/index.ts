export const getStringDate = (monthsAhead: number, baseDate?: Date) => {
  if(monthsAhead <= 0){
    throw new Error('Months ahead must be a positive number');
  }
  
  const targetDate = baseDate || new Date();
  
  const futureDate = targetDate.setMonth(targetDate.getMonth() + (monthsAhead - 1));

  const stringDate = new Intl.DateTimeFormat("en-US", {
    month: "long",
    year: "numeric",
  }).format(futureDate);

  return stringDate;
};
