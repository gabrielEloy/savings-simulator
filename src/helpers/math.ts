export const calculateMonthlyValue = (months: number, totalValue?: number) => totalValue ? totalValue/months : 0