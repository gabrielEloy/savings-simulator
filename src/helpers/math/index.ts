export const calculateMonthlyValue = (months: number, totalValue?: number) => {
    if(months <= 0){
        throw new TypeError('months must be a number bigger than zero');
    }

    return totalValue ? totalValue/months : 0
}