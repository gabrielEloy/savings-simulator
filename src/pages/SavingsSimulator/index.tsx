import { Header } from "components/Header";
import Subtitle from "components/typography/Subtitle";
import Wrapper from "AppStyles";
import SimulatorContainer from './components/SimulatorContainer';


interface Props {}

const App = (props: Props) => {
  return (
      <Wrapper>
        <Header />
        <Subtitle className="subtitle" color="#1B31A8">
          Let's plan your <strong>saving goal.</strong>
        </Subtitle>
        <div className="content">
         <SimulatorContainer />
        </div>
      </Wrapper>
  );
};

export default App;
