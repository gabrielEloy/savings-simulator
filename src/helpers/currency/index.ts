import { replaceAll } from 'helpers/string';
import INumberCurrencyToString from 'interfaces/INumberCurrencyToString';

export const stringCurrenctyToNumber = (value: string) => parseFloat(replaceAll(value, {',': ''}));

export const numberCurrencyToString = (value: number = 0, options: INumberCurrencyToString = {}) => {
    const { targetLanguage = 'en-US' }  = options;
    
    const localization = new Intl.NumberFormat(targetLanguage, { style: 'currency', currency: 'usd', ...options});
    return localization.format(value)
};