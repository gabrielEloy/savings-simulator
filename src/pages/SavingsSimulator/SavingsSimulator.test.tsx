import SavingsSimulator from "./index";
import { render, fireEvent, screen } from "@testing-library/react";
import { MONEY_GOAL, FUTURE_MONTHS } from "constants/Initialvalues";
import { numberCurrencyToString } from "helpers/currency";
import { getStringDate } from "helpers/dates";
import { calculateMonthlyValue } from "helpers/math";

describe("Savings simulator component", () => {
  describe("check app's initial structure", () => {
    it('should match the snapshot',  () => {
      const { container } = render(<SavingsSimulator />);

      expect(container).toMatchSnapshot();
    })

    describe("data inputs", () => {
      it("Should render the data input's with the expected initial values", () => {
        render(<SavingsSimulator />);

        const initialValueGoal = numberCurrencyToString(MONEY_GOAL, {
          style: "decimal",
        });
        const initialValue = screen.getByDisplayValue(initialValueGoal);

        const [month, year] = getStringDate(FUTURE_MONTHS).split(" ");
        const initialTargetMonth = screen.getByText(month);
        const initialTargetYear = screen.getByText(year);

        expect(initialValue).toBeTruthy();
        expect(initialTargetMonth).toBeTruthy();
        expect(initialTargetYear).toBeTruthy();
      });
    });
    describe("Monthly amount container", () => {
      it("Should have the expectrd initial monthly value", () => {
        render(<SavingsSimulator />);

        const expectedValue = calculateMonthlyValue(FUTURE_MONTHS, MONEY_GOAL);
        const formattedExpectedValue = numberCurrencyToString(expectedValue);

        const calculatedValue = screen.getAllByText(formattedExpectedValue);

        expect(calculatedValue).toBeTruthy();
        expect(calculatedValue.length).toEqual(2);
      });
    });
  });
  describe("Values update", () => {
    describe("Total amount", () => {
      it("Should correclty update the monthly amount when the total is changed", () => {
        render(<SavingsSimulator />);

        //querying the input by it's initial value
        const initialValueGoal = numberCurrencyToString(MONEY_GOAL, {
          style: "decimal",
        });
        const goalInput = screen.getByDisplayValue(initialValueGoal);

        const newGoal = 50000;

        fireEvent.change(goalInput, { target: { value: newGoal } });

        const expectedValue = calculateMonthlyValue(FUTURE_MONTHS, newGoal);
        const formattedExpectedValue = numberCurrencyToString(expectedValue);

        const newMonthlyValue = screen.getAllByText(formattedExpectedValue);

        expect(newMonthlyValue).toBeTruthy();
      });

      it("Should show a monthly value of 0 when the goal input has no value", () => {
        render(<SavingsSimulator />);

        //querying the input by it's initial value
        const initialValueGoal = numberCurrencyToString(MONEY_GOAL, {
          style: "decimal",
        });
        const goalInput = screen.getByDisplayValue(initialValueGoal);

        fireEvent.change(goalInput, { target: { value: "" } });

        const formattedExpectedValue = numberCurrencyToString(0);

        const newMonthlyValue = screen.getAllByText(formattedExpectedValue);

        expect(formattedExpectedValue).toEqual("$0.00");
        expect(newMonthlyValue).toBeTruthy();
      });
    });

    describe("Target month", () => {
      it("Should correclty update the monthly amount when the target month is updated to the future", () => {
        render(<SavingsSimulator />);

        const [nextMonth] = getStringDate(FUTURE_MONTHS + 1).split(" ");
        const [twoMonthsAhead] = getStringDate(FUTURE_MONTHS + 2).split(" ");

        const arrowRight = screen.getByLabelText("arrow-right");

        fireEvent.click(arrowRight);

        const initialTargetMonthQuery = screen.getByText(nextMonth);
        expect(initialTargetMonthQuery).toBeTruthy();

        fireEvent.click(arrowRight);

        const twoMonthsAheadQuery = screen.getByText(twoMonthsAhead);
        expect(twoMonthsAheadQuery).toBeTruthy();
      });
      it("Should correclty update the monthly amount when the target month is updated to the past", () => {
        render(<SavingsSimulator />);

        const [pastMonth] = getStringDate(FUTURE_MONTHS - 1).split(" ");
        const [twoMonthsAgo] = getStringDate(FUTURE_MONTHS - 2).split(" ");

        const arrowLeft = screen.getByLabelText("arrow-left");

        fireEvent.click(arrowLeft);

        const initialTargetMonthQuery = screen.getByText(pastMonth);
        expect(initialTargetMonthQuery).toBeTruthy();

        fireEvent.click(arrowLeft);

        const twoMonthsAgoQuery = screen.getByText(twoMonthsAgo);
        expect(twoMonthsAgoQuery).toBeTruthy();
      });
    });
  });
});
