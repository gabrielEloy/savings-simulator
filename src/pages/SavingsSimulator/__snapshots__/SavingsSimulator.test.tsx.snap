// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Savings simulator component check app's initial structure should match the snapshot 1`] = `
<div>
  <div
    class="sc-dlfnbm gpTfrs"
  >
    <div
      class="sc-bdfBwQ hLBLF"
    >
      <img
        alt="origin"
        src="brand_logo.svg"
      />
    </div>
    <h2
      class="sc-gsTCUz eNtpqv subtitle"
      color="#1B31A8"
    >
      Let's plan your 
      <strong>
        saving goal.
      </strong>
    </h2>
    <div
      class="content"
    >
      <div
        class="sc-hKgILt eCAnQM"
      >
        <div
          class="simulator-container"
        >
          <div
            class="title"
          >
            <img
              alt="house icon"
              src="house.svg"
            />
            <div
              class="text"
            >
              <h3
                class="sc-jSgupP fegsyl header-text"
              >
                Buy a house
              </h3>
              <p
                class="sc-gKsewC hdTzWd"
              >
                Saving goal
              </p>
            </div>
          </div>
          <div
            class="data-inputs"
          >
            <div
              class="sc-fubCfw gHrCnJ"
            >
              <span
                class="sc-iBPRYJ fNbmqV label"
              >
                Total amount
              </span>
              <div
                class="input-container"
              >
                <div
                  class="preffix-container"
                >
                  <img
                    alt="Dollar sign"
                    src="dollar-sign.svg"
                  />
                </div>
                <input
                  class="total"
                  placeholder="insert the desired value"
                  value="25000"
                />
              </div>
            </div>
            <div
              class="sc-pFZIQ eZNIEq months"
              tabindex="0"
            >
              <div
                class="label"
              >
                <span
                  class="sc-iBPRYJ fNbmqV label"
                >
                  Reach goal by
                </span>
              </div>
              <div
                class="date-selector-content"
              >
                <button
                  class="controls"
                >
                  <svg
                    aria-label="arrow-left"
                    fill="none"
                    height="24"
                    name="chevron-left"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M15.707 5.293a1 1 0 010 1.414L10.414 12l5.293 5.293a1 1 0 01-1.414 1.414l-6-6a1 1 0 010-1.414l6-6a1 1 0 011.414 0z"
                      fill="#8A9CA9"
                      fill-rule="evenodd"
                    />
                  </svg>
                </button>
                <div
                  class="input"
                >
                  <p
                    class="sc-gKsewC iWaxtf"
                    color="#1E2A32"
                    font-weight="semi-bold"
                  >
                    December
                  </p>
                  <p
                    class="sc-gKsewC hdTzWd"
                  >
                    2021
                  </p>
                </div>
                <button
                  class="controls"
                >
                  <svg
                    aria-label="arrow-right"
                    fill="none"
                    height="24"
                    name="chevron-left"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M8.293 5.293a1 1 0 011.414 0l6 6a1 1 0 010 1.414l-6 6a1 1 0 01-1.414-1.414L13.586 12 8.293 6.707a1 1 0 010-1.414z"
                      fill="#8A9CA9"
                      fill-rule="evenodd"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
          <div
            class="montly-amout-container"
          >
            <div
              class="values"
            >
              <h2
                class="sc-gsTCUz jwvVTA"
              >
                Monthly amount
              </h2>
              <h3
                class="sc-jSgupP jSoLBx"
                color="#0079FF"
              >
                $2,500.00
              </h3>
            </div>
            <div
              class="desc"
            >
              <span
                class="sc-eCssSg lfaeVv"
              >
                Youâ€™re planning 
                <strong>
                  10
                   monthly deposits
                </strong>
                 to reach your 
                <strong>
                  $2,500.00
                </strong>
                 goal by
                 
                <strong>
                  December 2021
                </strong>
              </span>
            </div>
          </div>
          <div
            class="confirm-button-container"
          >
            <button
              class="sc-jrAGrp gfAeDP"
            >
              Confirm
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
